### PR TITLE
Add REL bin plugin

### LINKED ISSUE #20508: Support REL binaries
## Description

https://wiki.tockdom.com/wiki/REL_(File_Format)

REL is a relocatable library format used by Nintendo Wii

### PR DESCRIPTION
**Pull request by terorie**
- [x] Mark this if you consider it ready to merge
- [x] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

**Description**

https://wiki.tockdom.com/wiki/REL_(File_Format)

- [x] Section loading
- [x] Meson build
- [x] Make build
- [ ] Profiling / performance
- [x] Fuzzing
- [x] Internal relocs
- [x] External relocs to ROM/symbols
- [x] ~~External relocs to PIC/imports~~ (out of scope)
- [ ] Test DOL/REL linking
- [x] Add tests
- [x] Formatting, code cleanup

Closes #20508

### CODE REVIEW

**Comment by trufae** [comment]
[r] Cannot open '/Users/richard/prj/mkw/artifacts/orig/pal/StaticR.rel'
[53](https://github.com/radareorg/radare2/runs/7598718852?check_suite_focus=true#step:8:54)

send this file to this repo https://github.com/radareorg/radare2-testbins/pulls

File: libr/bin/p/bin_rel.c [hunk 1]
```diff
@@ -0,0 +1,613 @@
+/* radare - LGPL - 2022 - terorie */
+
+#include <r_types.h>
+#include <r_util.h>
+#include <r_lib.h>
+#include <r_bin.h>
+#include <string.h>
+
+#include "../format/elf/glibc_elf.h"
```
Comment (terorie) [inline libr/bin/p/bin_rel.c:9]: no need to have this here

File: test/db/formats/rel [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+pd10 @0x8055531c
+EOF
+EXPECT=<<EOF
+            ;-- rel1_prolog:
+            0x8055531c      9421fff0       stwu r1, -0x10(r1)
+            0x80555320      7c0802a6       mflr r0
+            0x80555324      90010014       stw r0, 0x14(r1)
+            0x80555328      93e1000c       stw r31, 0xc(r1)
+            0x8055532c      3fe00000       lis r31, 0
+            0x80555330      3bff0000       addi r31, r31, 0
```
Comment (terorie) [inline test/db/formats/rel:14]: relocs are completely missing here

Comment (terorie) [reply test/db/formats/rel]: fixed with `-e io.cache=true`

**Comment by terorie** [comment]
Ready for review, pending merge of https://github.com/radareorg/radare2-testbins/pull/78

**Comment by terorie** [comment]
libFuzzer found something 🗡️ 

```
% r2 -F rel oom-3852cb7a9012f8d9f2f3c21ad599c97b1caee477
INFO: REL module ID 00000003
INFO: REL version 3

zsh: segmentation fault  r2 -F rel oom-3852cb7a9012f8d9f2f3c21ad599c97b1caee477
```

**Comment by terorie** [comment]
Fixed OOM fuzz findings by adding limits to section/imp/reloc counts. Will let libFuzzer run for another hour or two

File: test/db/formats/rel [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+INFO: REL version 3
+DEBUG: imp 0: 2 relocs
+DEBUG: imp 1: 2 relocs
+DEBUG: imp 2: 2 relocs
+DEBUG: imp 3: 2 relocs
+DEBUG: imp 4: 2 relocs
+DEBUG: imp 5: 2 relocs
+DEBUG: imp 6: 7 relocs
+DEBUG: imp 7: 16 relocs
+ERROR: Imports from other REL (00000002) not yet implemented
```
Comment (trufae) [inline test/db/formats/rel:45]: this is a warning, if it was an error. it will not report anything after this message.

Comment (trufae) [reply test/db/formats/rel:45]: lets say warning or debug or info, but not error

**Review by trufae — CHANGES_REQUESTED** [review]
wonder why the magic search is failing in this pr 🤔

**Comment by trufae** [comment]
make it grin and ill merge

**Comment by trufae** [comment]
ping

**Comment by trufae** [comment]
Tired of waiting, i just fixed the crash and merged the thing. thanks!!

**Comment by terorie** [comment]
Apologies for the wait 😢 Will try to make it up with some backlogged tickets

**Comment by trufae** [comment]
Np :D thanks for the contrib! It was just a tiny thing to get it in. I know its tedious to keep up with all the prs. So its ok for me to take over them

**Pull request closed without merging**